### PR TITLE
Make topics more extendable and fix topicsToFcm

### PR DESCRIPTION
**Pull request by Stevemoretz**
A simple pull request to change a few private methods to protected.
In my use case I'm extending this class and it's annoying that these methods are private.

### CODE REVIEW

**Review by williamdes — APPROVED** [review]
💯 
This looks good, could you let me know more about what you would need this lib to do for you ?

**Comment by Stevemoretz** [comment]
> 💯 This looks good, could you let me know more about what you would need this lib to do for you ?

Thank you, sure!

Long story:
I'm building a push notification manager system (Laravel uses this for everything it creates that has drivers), I'm using this library internally in my own library which makes FCM one of my drivers, long story short, topics can be used in other drivers that I have as well, for the sake of consistency I have to extend the topics class from this library and make it work with my other drivers, but the Topics class only has private methods, so I had to make the necessary method protected so they can be changed, and I can use the same syntax for all my drivers.

if you are interested:
in the extended class for other drivers, I use eloquent to build a query which selects the topics from the database, instead of building the topics string.

Long story shorter : 
Topics class is not extendable at all, everything that is not public is private, which is a little wrong to me even if I didn't have my use case.

**Comment by Stevemoretz** [comment]
By the way, in the last days I've been pushing a lot of pull request around the github and almost nobody cared enough about their library to even come and take a look, thank you for caring about your library.
I'm going to push a fix to another issue that this library has as well, which I noticed a long time ago.

**Comment by Stevemoretz** [comment]
This new commit:
https://github.com/code-lts/Laravel-FCM/pull/17/commits/78d5a084b50dfe923d61dbdda382caf064167714

Fixes a problem that happened to me a long time ago, there was no errors and it simply wasn't working the way I intended, here's one example:

```php
        $condition = (new PNTopics())->topic('helo')->nest(function (PNTopics $item){
            return $item->topic('s')->orTopic('s2');
        },"&&")->build()['condition'];
```

the output of this is:

```
'helo' in topics&&('s' in topics || 's2' in topics)
```

since there is no space between && and other characters, this causes a bug without any errors.
with the new commit, it would output this:

```
'helo' in topics && ('s' in topics || 's2' in topics)
```
And if you use anything other than || and && it would throw an error.

```php
        $condition = (new PNTopics())->topic('helo')->nest(function (PNTopics $item){
            return $item->topic('s')->orTopic('s2');
        }," && ")->build()['condition'];
```

would give :
Exception : You must use either one of (|| , &&) as a condition

However now I notice this is a breaking change, so I'll add another commit to add backward compatibility!

**Comment by Stevemoretz** [comment]
Fixed, now works with '||' and '&&' and ' || ' and ' && ' anything else throws the error.

File: src/Message/Topics.php [hunk 1]
```diff
@@ -137,9 +137,25 @@ public function nest(Closure $callback, $condition)
...
+                throw new \Exception(
+                    "You must use either one of (|| , &&) as a condition"
+                );
+            }
 
             $this->conditions[] = [
-                'condition' => $condition,
+                "condition" => $needsBackwardCompatibility
+                    ? $condition
+                    : sprintf(" %s ", $condition),
```
Comment (williamdes) [inline src/Message/Topics.php]: Could you add some tests for this please ?

**Comment by williamdes** [comment]
Thanks a lot for your nice messages !
I try to do my best for this lib I adopted into @code-lts some years ago

Your fix looks really good, I am a bit hesitant on the use of "collect()" but anyway, if there is tests for it this should be okay

**Comment by Stevemoretz** [comment]
Thank you so much yeah sure!

```php
<?php

namespace Push;

use LaravelFCM\Message\Topics;

class TopicsFixTest extends \TestCase {
    private function getResultTest($condition){
        return (new Topics())->topic('helo')->nest(function (Topics $item){
            return $item->topic('s')->orTopic('s2');
        },$condition)->build()['condition'];
    }
    /**
     * A basic feature test example.
     *
     * @return void
     */
    public function testExample() {
        self::assertEquals($this->getResultTest('&&'),$this->getResultTest(' && '));
        self::assertEquals($this->getResultTest('||'),$this->getResultTest(' || '));
    }
}
```

File: src/Message/Topics.php [hunk 2]
```diff
@@ -137,9 +137,25 @@ public function nest(Closure $callback, $condition)
...
+            if (
+                !$needsBackwardCompatibility &&
+                !collect()
+                    ->push("||", "&&")
+                    ->contains(null, $condition)
+            ) {
+                throw new \Exception(
+                    "You must use either one of (|| , &&) as a condition"
+                );
+            }
```
Comment (williamdes) [inline src/Message/Topics.php]: Could this be written in a more PHP pure solution ?

Thanks for the tests !

Comment (Stevemoretz) [reply src/Message/Topics.php]: Sure I'll send a pure implementation soon, no problem at all!

Comment (Stevemoretz) [reply src/Message/Topics.php]: Done here : https://github.com/code-lts/Laravel-FCM/pull/17/commits/eb001e7695b00a62878dee775843a0d507518e2c

The same tests above still work.

**Review by williamdes — APPROVED** [review]
I love it!
And understand it much better

**Comment by Stevemoretz** [comment]
> I love it! And understand it much better

Thank you so much! That's kind of you

**Review by williamdes — APPROVED** [review]

**Comment by williamdes** [comment]
What do you think about it now @Stevemoretz ?

Looks like it would break no workflow and still allow some hacky users to put other operators if they want to like I changed it
Also I only changed the formatted so the data is not altered before formatting into a string

**Comment by Stevemoretz** [comment]
> What do you think about it now @Stevemoretz ?
> 
> Looks like it would break no workflow and still allow some hacky users to put other operators if they want to like I changed it Also I only changed the formatted so the data is not altered before formatting into a string

Well the test sure looks a lot better now, but since you remove the checks and exceptions, if users put 'and' and 'or' it would silently fail, I think it was better throwing the exception but if there is anything else a user can pass in, I'd suggest a third argument to the `nest` function which could be `$strict = true` and if the user passes false it can then pass hacky stuff (by not running the checks and throwing the exception or course) otherwise he/she won't make no mistakes.

**Comment by williamdes** [comment]
> > What do you think about it now @Stevemoretz ?
> > Looks like it would break no workflow and still allow some hacky users to put other operators if they want to like I changed it Also I only changed the formatted so the data is not altered before formatting into a string
> 
> Well the test sure looks a lot better now, but since you remove the checks and exceptions, if users put 'and' and 'or' it would silently fail, I think it was better throwing the exception but if there is anything else a user can pass in, I'd suggest a third argument to the `nest` function which could be `$strict = true` and if the user passes false it can then pass hacky stuff (by not running the checks and throwing the exception or course) otherwise he/she won't make no mistakes.

Very good idea, I added back your work and added tests for this solution !
Merging this, let me know if it looks good on your end

**Comment by Stevemoretz** [comment]
Hi thank you for your support!
It looks fantastic, and I learnt a couple of things from your unit tests, thank you!

**Comment by williamdes** [comment]
> Hi thank you for your support! It looks fantastic, and I learnt a couple of things from your unit tests, thank you!

Hi @Stevemoretz !
Your work is now live !
[1.7.0](https://github.com/code-lts/Laravel-FCM/releases/tag/1.7.0) is out 🚀

**Comment by Stevemoretz** [comment]
Thank you so much for informing me!

**Pull request merged**